### ISSUE TITLE
 sct_segment_graymatter crazy results with some newer version of python packages

### ISSUE THREAD

**Issue by zougloub** [neuropoly/spinalcordtoolbox#1651]
Similar to #1619, but for sct_segment_graymatter on `sct_example_data/mt` (sct_testing is OK).

**Comment by zougloub** [comment]
Performing `conda install numpy`, which performs the following updates on amd64:
```
    cairo:                         1.14.8-0               --> 1.14.12-h77bcde2_0   
    conda:                         4.3.30-py27h6ae6dc7_0  --> 4.5.0-py27_0         
    conda-env:                     2.6.0-0                --> 2.6.0-h36134e3_1     
    fontconfig:                    2.12.1-3               --> 2.12.4-h88586e7_1    
    freetype:                      2.5.5-2                --> 2.8-hab7d2ae_1       
    glib:                          2.50.2-1               --> 2.53.6-h5d9569c_2    
    harfbuzz:                      0.9.39-2               --> 1.7.4-hc5b324e_0     
    libpng:                        1.6.30-1               --> 1.6.34-hb9fc6fc_0    
    libtiff:                       4.0.6-3                --> 4.0.9-h28f6b97_0     
    matplotlib:                    1.5.1-np111py27_0      --> 2.2.2-py27h0e671d2_0 
    mkl:                           11.3.3-0               --> 2018.0.1-h19d6760_4  
    numpy:                         1.11.2-py27_0          --> 1.14.2-py27hdbf6ddf_0
    pandas:                        0.18.1-np111py27_0     --> 0.22.0-py27hf484d3e_0
    pango:                         1.40.3-1               --> 1.41.0-hd475d92_0    
    pillow:                        4.2.1-py27_0           --> 5.0.0-py27h3deb7b8_0 
    pycosat:                       0.6.1-py27_1           --> 0.6.3-py27ha4109ae_0 
    pyqt:                          4.11.4-py27_4          --> 5.6.0-py27h4b1e83c_5 
    python:                        2.7.12-1               --> 2.7.14-h9b67528_20   
    qt:                            4.8.7-4                --> 5.6.2-h974d657_12    
    readline:                      6.2-2                  --> 7.0-ha6073c6_4       
    scikit-image:                  0.12.3-np111py27_1     --> 0.13.1-py27h14c3975_1
    scikit-learn:                  0.17.1-np111py27_2     --> 0.19.1-py27h445a80a_0
    scipy:                         0.18.1-np111py27_0     --> 1.0.0-py27hf5f0f52_0 
    sqlite:                        3.13.0-0               --> 3.22.0-h1bed415_0    
    tk:                            8.5.18-0               --> 8.6.7-hc745277_3     
    zlib:                          1.2.8-3                --> 1.2.11-ha838bed_2    
```
yields the problem.

Trying to narrow it down...

**Comment by zougloub** [comment]
Upgrading numpy alone -> no problem.

**Comment by zougloub** [comment]
Problem caused by `scikit-learn: 0.17.1-np111py27_2 --> 0.19.1-py27h445a80a_0`

Actually there's a warning printed:
```UserWarning: Trying to unpickle estimator PCA from version pre-0.18 when using version 0.19.1. This might lead to breaking code or invalid results. Use at your own risk.
```

-> Linking to #1616

**Comment by jcohenadad** [comment]
I think this deserves high priority (i.e., if some of our users get wrong segmentation results, this has critical implications).

**Comment by jcohenadad** [comment]
@SaraDupont can you please indicate (or redirect to documentation) on how to generate a new model? As you see the function is currently failing due to recent updates, and we absolutely need it before the next SCT course. Thank you :-)

**Comment by SaraDupont** [comment]
There's a small text at the beginning of `sct_segment_graymatter` (copied below) and the model generation is done by `msct_multiatlas_seg`. The default parameters should be the ones that gave the best results on the 80 subjects dataset (most of the parameters should be listed in the info text file of the model if you want to make sure what was used in the previous version) 

############################################################
INFORMATION:
The model used in this function is compound of:
  - a dictionary: a list of slices of WM/GM contrasted images with their manual segmentations [slices.pklz]
  - a model representing this dictionary in a reduced space (a PCA or an isomap model as implemented in sk-learn) [fitted_model.pklz]
  - the dictionary data fitted to this model (i.e. in the model space) [fitted_data.pklz]
  - the averaged median intensity in the white and gray matter in the model [intensities.pklz]
  - an information file indicating which parameters were used to construct this model, and te date of computation [info.txt]

A constructed model is provided in the toolbox here: `$PATH_SCT/data/gm_model`.
It's made from T2* images of 80 subjects and computed with the parameters that gives the best gray matter segmentation results.
However you can compute you own model with your own data or with other parameters and use it to segment gray matter by using  the flag `-model path_new_gm_model/`.
To do so, you should have a folder (path_to_dataset/) containing for each subject (with a folder per subject):
        - a WM/GM contrasted image (for ex T2*-w) containing 'im' in its name
        - a segmentation of the spinal cord containing 'seg' in its name
        - a (or several) manual segmentation(s) of the gray matter containing 'gm' in its(their) name(s)
        - a level file containing 'level' in its name : it can be an image containing a level label per slice indicating at wich vertebral level correspond this slice (usually obtained by registering the PAM50 template to the WM/GM contrasted image) or a text file indicating the level of each slice.

For more information on the parameters available to compute the model, type:
`msct_multiatlas_seg -h`

to compute the model, use the following command line :
`msct_multiatlas_seg -path-data path_to_dataset/`
Then use the folder `gm_model/` (output from msct_multiatlas_seg) in this function with the flag `-model gm_model/`

**Comment by jcohenadad** [comment]
@zougloub maybe a quicker (but temporary) workaround: can you try to force back scikit-learn==0.17.1 after numpy update on your station and see if it solves the issue?

**Comment by jcohenadad** [comment]
@zougloub any update on this? for a quick fix, and given that `sct_segment_graymatter` will eventually be replaced by `sct_deepseg_gm`, should we just go with the "quick fix" of downgrading sklearn?

**Comment by zougloub** [comment]
@jcohenadad I think a better workaround is to bump the sklearn version and to regenerate the model data before the next release.

**Comment by jcohenadad** [comment]
ok, but i expect this will probably take several hours (if not days). I don't have time to do this now and it requires in situ access to the data. I am hesitating between investing time in this and dropping this function in the next releases (considering `sct_deepseg_gm` performs better).

**Comment by joshuacwnewton** [comment]
> I am hesitating between investing time in this and dropping this function in the next releases (considering sct_deepseg_gm performs better).

This is exactly what happened! `sct_segment_graymatter` is no more thanks to #2549. Closing. :)
